nix/master.nix

```diff
@@ -492,6 +492,38 @@ in
         '';
         default = 50;
       };
+
+      branches = {
+        build_prs = lib.mkOption {
+          type = lib.types.bool;
+          description = ''
+            When enabled buildbot-nix will schedule builds for every PR opened.
+            It will however not register GC roots.
+          '';
+          default = true;
+        };
+
+        primary = lib.mkOption {
+          type = lib.types.str;
+          description = ''
+            A `re` Python compatible regex specifying which additional branches should
+            buildbot-nix consider in addition to the default branch. Like for the default
+            branch, buildbot-nix will register GC roots for any builds it performs for this
+            branch. An empty regex means to not consider any additional branches.
+          '';
+          default = "";
+        };
+
+        secondary = lib.mkOption {
```

Comment:
Do we need to differentiate here? Can we not generate a GC root for all explicitly defined branches?

Reply:
We could, thats another option

Reply:
maybe the easier thing to do for now is to be less opinionated and expose all these knobs in the config. `branches` could be a list of attrs.
because i'm not that well versed on top of my head with the type library, here's a pseudo-code type for the `options.branches`:

`[ { branchRe: str, buildPRs: bool, registerGCRoots: bool, copyOutputs: bool ]`

this gives enough flexibility for all discussed use-cases.

Reply:
So a list of regex selectors and then flags on each one, dictating what to do

Reply:
yes, that's the idea. it could also be an attrset in which the name is a unique identifier instead of anonymous attrs in a list. i've found lists harder to debug.

Reply:
It'll be in TOML maybe, since im looking into making this per-repo. Though maybe a default list + per-repo would be good

Reply:
example

```nix
branches.release = {
   match = "release-.*";
   buildPRs = true;
   registerGCRoots = true;
   copyOutputs = true;
};
```

Reply:
``
[branches.release]
match = "release-*"
buildsPRs = true
registerGCRoots = true
copyOutputs = true
```

In a `buildbot-nix.toml` file in the root of the default branch

Reply:
> It'll be in TOML maybe, since im looking into making this per-repo. Though maybe a default list + per-repo would be good

i'm a bit wary of attempting per-repo settings in the first iteration because it opens up a lot of security questions. i agree that it's a nice addon for a subsequent iteration.

Reply:
> In a `buildbot-nix.toml` file in the root of the default branch

i assume it's possible to transform the nix expression into toml and feed that to the python implementation handling this code that part can be implemented and later be reused for files from the repo.

Reply:
Well it'll be largely the same code. As in the per-repo `BranchConfig` object gets merged with the global one, *somehow*. So technically the only security critical part is tbe *somehow*. But yeah probanly. I'll finish up #318 without being able to declare branches

Reply:
I'm proceeding with this design, except that `buildPRs` cannot be implemented the way I took it which is "build PRs based on which branch they're based on, and if that branch matches a glob". Such an implementation is not possible as buildbot doesn't have nor expose this information for a `Change` (see [Change](https://github.com/buildbot/buildbot/blob/master/master/buildbot/changes/changes.py#L33) and [ChangeFilter](https://github.com/buildbot/buildbot/blob/5e9c38f8948f622521420a25f5f250d53f273fe2/master/buildbot/changes/filter.py#L26)). So rather than that I think we just keep `buildPRs` as a instance global option (repository scoped once that's in) and keep the simple behavior with PRs.